tokio/src/task/scope/intrusive_double_linked_list.rs

```diff
@@ -0,0 +1,628 @@
+        let tail = self.tail;
+        self.head = null_mut();
+        self.tail = null_mut();
+        LinkedList::<T> { head, tail }
+    }
+
+    /// Returns whether the linked list doesn not contain any node
+    #[allow(dead_code)]
+    pub(crate) fn is_empty(&self) -> bool {
+        if !self.head.is_null() {
+            return false;
+        }
+
+        assert!(self.tail.is_null());
+        true
+    }
+
+    /// Removes the given item from the linked list.
+    /// Returns whether the item was removed.
+    /// The function is only safe as long as valid pointers are stored inside
+    /// the linked list.
+    pub(crate) unsafe fn remove(&mut self, item: *mut ListNode<T>) -> bool {
+        if item.is_null() {
+            return false;
+        }
+
+        let prev = (*item).prev;
+        if prev.is_null() {
+            // This might be the first item in the list
+            if self.head != item {
+                return false;
+            }
+            self.head = (*item).next;
+        } else {
+            debug_assert_eq!((*prev).next, item);
+            (*prev).next = (*item).next;
+        }
+
+        let next = (*item).next;
+        if next.is_null() {
+            // This might be the last item in the list
+            if self.tail != item {
+                return false;
+            }
+            self.tail = (*item).prev;
+        } else {
+            debug_assert_eq!((*next).prev, item);
+            (*next).prev = (*item).prev;
+        }
+
+        (*item).next = null_mut();
+        (*item).prev = null_mut();
+
+        true
+    }
+}
+
+/// An iterator over an intrusively linked list
+pub(crate) struct LinkedListIterator<T> {
+    current: *mut ListNode<T>,
+}
+
+impl<T> Iterator for LinkedListIterator<T> {
+    type Item = *mut ListNode<T>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.current.is_null() {
+            return None;
+        }
+
+        let node = self.current;
+        // Safety: This is safe as long as the linked list is intact, which was
+        // already required through the unsafe when creating the iterator.
+        unsafe {
+            self.current = (*self.current).next;
+        }
+        Some(node)
+    }
+}
+
+/// An iterator over an intrusively linked list
+pub(crate) struct LinkedListReverseIterator<T> {
+    current: *mut ListNode<T>,
+}
+
+impl<T> Iterator for LinkedListReverseIterator<T> {
+    type Item = *mut ListNode<T>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.current.is_null() {
+            return None;
+        }
+
+        let node = self.current;
+        // Safety: This is safe as long as the linked list is intact, which was
+        // already required through the unsafe when creating the iterator.
+        unsafe {
+            self.current = (*self.current).prev;
+        }
+        Some(node)
+    }
+}
+
+#[cfg(test)]
+#[cfg(feature = "std")] // Tests make use of Vec at the moment
```

Comment:
Feature `std` doesn't exist (yet), so can't run tests. I'm guessing probably just remove this?